renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -2894,6 +3138,99 @@ struct VulkanPixelHistoryPerFragmentCallback : VulkanPixelHistoryCallback
+
+          clearAtts[0].aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+          clearAtts[0].colorAttachment = colorOutputIndex;
+          memcpy(clearAtts[0].clearValue.color.float32, premod.col.floatValue.data(),
+                 sizeof(clearAtts[0].clearValue.color));
+
+          clearAtts[1].aspectMask = VK_IMAGE_ASPECT_DEPTH_BIT;
+          clearAtts[1].clearValue.depthStencil.depth = premod.depth;
+
+          if(IsDepthOrStencilFormat(m_CallbackInfo.targetImageFormat))
+            ObjDisp(cmd)->CmdClearAttachments(Unwrap(cmd), 1, clearAtts + 1, 1, &rect);
+          else
+            ObjDisp(cmd)->CmdClearAttachments(Unwrap(cmd), 2, clearAtts, 1, &rect);
+
+          // Note: pipes.redrawToBeforePipe has the stencil compare op set to VK_COMPARE_OP_GREATER,
+          // so this reference is not inclusive
+          ObjDisp(cmd)->CmdSetStencilCompareMask(Unwrap(cmd), VK_STENCIL_FACE_FRONT_AND_BACK, 0xff);
+          ObjDisp(cmd)->CmdSetStencilWriteMask(Unwrap(cmd), VK_STENCIL_FACE_FRONT_AND_BACK, 0xff);
+          ObjDisp(cmd)->CmdSetStencilReference(Unwrap(cmd), VK_STENCIL_FACE_FRONT_AND_BACK, f);
+
+          const ActionDescription *action = m_pDriver->GetAction(eid);
+          m_pDriver->ReplayDraw(cmd, *action);
+
+          // Now actually draw the current one. Stencil op is set to VK_COMPARE_OP_EQUAL.
+          VkMarkerRegion::Set(StringFormat::Fmt("Getting blend %s for fragment %u in %u",
+                                                i == 0 ? "source" : "destination", f, eid),
+                              cmd);
+
+          state.graphics.pipeline = GetResID(i == 0 ? pipes.blendSrcPipe : pipes.blendDstPipe);
+          state.BindPipeline(m_pDriver, cmd, VulkanRenderState::BindGraphics, false);
+
+          ObjDisp(cmd)->CmdClearAttachments(Unwrap(cmd), 1, &stencilAtt, 1, &rect);
+          ObjDisp(cmd)->CmdSetStencilCompareMask(Unwrap(cmd), VK_STENCIL_FACE_FRONT_AND_BACK, 0xff);
+          ObjDisp(cmd)->CmdSetStencilWriteMask(Unwrap(cmd), VK_STENCIL_FACE_FRONT_AND_BACK, 0xff);
+          ObjDisp(cmd)->CmdSetStencilReference(Unwrap(cmd), VK_STENCIL_FACE_FRONT_AND_BACK, f);
+
+          m_pDriver->ReplayDraw(cmd, *action);
+          state.EndRenderPass(cmd);
+
+          CopyImagePixel(cmd, colourCopyParams,
+                         (fragsProcessed + f) * sizeof(PerFragmentInfo) + offset);
+
+          // TODO: is this useful?
+          if(depthImage != VK_NULL_HANDLE)
+          {
+            VkCopyPixelParams depthCopyParams = colourCopyParams;
+            depthCopyParams.srcImage = depthImage;
+            depthCopyParams.srcImageLayout = depthLayout;
+            depthCopyParams.srcImageFormat = depthFormat;
+            CopyImagePixel(cmd, depthCopyParams, (fragsProcessed + f) * sizeof(PerFragmentInfo) +
+                                                     offset +
+                                                     offsetof(struct PixelHistoryValue, depth));
+          }
```

Comment:
I collect depth information for all of these columns by reading the depth value each time. This seems kinda redundant, but the alternative would be to not use `PixelHistoryValue`/`ModificationValue` for this purpose, which seems equally messy (especially since `ModificationValue` is what has `IsValid()`, not `PixelValue`).